docs/qos_enforcement.md

```diff
@@ -0,0 +1,292 @@
+POST    /lte/{network_id}/qos_profiles
+
+GET     /lte/{network_id}/qos_profiles/{qos_name}
+PUT     /lte/{network_id}/qos_profiles/{qos_name}
+DELETE  /lte/{network_id}/qos_profiles/{qos_name}
+
+GET     /feg_lte/{network_id}/qos_profiles
+POST    /feg_lte/{network_id}/qos_profiles
+
+GET     /feg_lte/{network_id}/qos_profiles/{qos_name}
+PUT     /feg_lte/{network_id}/qos_profiles/{qos_name}
+DELETE  /feg_lte/{network_id}/qos_profiles/{qos_name}
+
+
+
+APN Profiles
+
+GET     /lte/{network_id}/apns
+POST    /lte/{network_id}/apns
+
+GET     /lte/{network_id}/apns/{apn_name}
+PUT     /lte/{network_id}/apns/{apn_name}
+DELETE  /lte/{network_id}/apns/{apn_name}
+
+
+
+APN Resources
+
+GET     /lte/{network_id}/gateways/{gateway_id}/resource_labels
+POST    /lte/{network_id}/gateways/{gateway_id}/resource_labels
+
+GET     /lte/{network_id}/gateways/{gateway_id}/resource_labels/{resource_label}
+PUT     /lte/{network_id}/gateways/{gateway_id}/resource_labels/{resource_label}
+DELETE  /lte/{network_id}/gateways/{gateway_id}/resource_labels/{resource_label}
+```
+
+
+
+### API Entity Definitions
+
+Entity definitions for QoS, APN, and subscribers will change. 
+The final state is summarized here:
+
+```
+qos_profile:
+  id: string
+  class_id: enum[0,255]
+  max_req_bw_dl: uint32
+  max_req_bw_ul: uint32
+  # Guaranteed bit rate
+  # Not enforced by Magma yet
+  gbr:
+    gbr_ul: uint32
+    gbr_dl: uint32
+  # Allocation and retention priority
+  # Not enforced by Magma yet
+  arp:
+    arp_priority_level: uint32
+    arp_preemption_capability: bool
+    arp_preemption_vulnerability: bool
+ # Aggregated Maximum bit rate
+ # for non-GBR bearers, this limits the aggregate bit rate
+ # Enforce this on pipelined
+ apn_ambr:
+   apn_ambr_dl: uint32
+   apn_ambr_ul: uint32
+```
+
+As an initial goal, we would like to support apn_ambr.
+
+```
+apn_configuration:
+  default_qos_profile: string
+  supported_qos_profiles: array[string]
+  # resource_label should match the name of a specified ApnResource
+  resource_label: string
+```
+
+Multiple QoS objects will be allowed to be associated to APN objects.
+
+One can be specified as the APN-level default policy. This is checked by 
+verifying a non-zero value for fields *apn_ambr_dl* and *apn_ambr_ul* for the 
+default QoS object.
+
+Additional QoS objects can be associated to the APN for additional resources 
+beyond the default policy, both physical and logical. 
+For fields *apn_ambr_dl* and *apn_ambr_ul*, these field values must be zero.
+
+```
+apn_resource:
+  id: string
+  # Determines the IPs that will be allocated to connected subscribers under
+  # this APN resource
+  ipv4_block: string 
+  # Specify either VLAN or the interface name for SGi
+  # VLAN to use when communicating over SGi interface
+  vlan_id: string
+  # uplink_bridge must be configured with the uplink interface for this value 
+  # to be valid.
+  uplink_interfaces: string
+  # Prefix for ipv6 allocation
+  ipv6_prefix: string
```

Comment:
Do we model an IPAddress object with both v6 and v4 also what happens during attach we return both the v6 and v4 address?

Reply:
We need to include PDNType in the APN Configuration. If it is set as IPV4V6, then both IP addresses are assigned and returned (to be completed in Q3) during attach.